app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt

```diff
@@ -0,0 +1,172 @@
+                                Prefs.setPushNotificationTokenOld("")
+                            }, {
+                                L.e(it)
+                                if (it is MwException && it.error.title == "echo-push-token-not-found") {
+                                    // token was not found in the database, so consider it gone.
+                                    Prefs.setPushNotificationTokenOld("")
+                                }
+                            })
+                } else {
+                    Prefs.setPushNotificationTokenOld("")
+                }
+            }
+
+            ServiceFactory.get(WikipediaApp.getInstance().wikiSite).subscribePush(csrfToken, token)
+                    .subscribeOn(Schedulers.io())
+                    .observeOn(AndroidSchedulers.mainThread())
+                    .retry(SUBSCRIBE_RETRY_COUNT.toLong())
+                    .subscribe({
+                        L.d("Token subscribed successfully.")
+                        Prefs.setPushNotificationTokenSubscribed(true)
+                    }, {
+                        L.e(it)
+                        if (it is MwException && it.error.title == "echo-push-token-exists") {
+                            // token already exists in the database, so consider it subscribed.
+                            Prefs.setPushNotificationTokenSubscribed(true)
+                        }
+                    })
+        }
+
+        private fun setNotificationOptions(csrfToken: String) {
+            val optionList = ArrayList<String>()
+
+            optionList.add("echo-subscriptions-push-edit-user-talk=" + if (Prefs.notificationUserTalkEnabled()) "1" else "0")
+            optionList.add("echo-subscriptions-push-login-fail=" + if (Prefs.notificationLoginFailEnabled()) "1" else "0")
+            optionList.add("echo-subscriptions-push-mention=" + if (Prefs.notificationMentionEnabled()) "1" else "0")
+            optionList.add("echo-subscriptions-push-thank-you-edit=" + if (Prefs.notificationMilestoneEnabled()) "1" else "0")
+            optionList.add("echo-subscriptions-push-reverted=" + if (Prefs.notificationRevertEnabled()) "1" else "0")
+            optionList.add("echo-subscriptions-push-edit-thank=" + if (Prefs.notificationThanksEnabled()) "1" else "0")
+            // Explicitly enable cross-wiki notifications
+            optionList.add("echo-cross-wiki-notifications=1")
+
+            ServiceFactory.get(WikipediaApp.getInstance().wikiSite).postSetOptions(optionList.joinToString(separator = "|"), csrfToken)
+                    .subscribeOn(Schedulers.io())
+                    .observeOn(AndroidSchedulers.mainThread())
+                    .subscribe({
+                        L.d("Notification options successfully.")
+                    }, {
+                        L.e(it)
+                    })
+        }
+
+        fun unsubscribePush(csrfToken: String): Observable<MwQueryResponse> {
```

Comment:
Would it be possible to reuse this method in `subscribeWithCsrf` to unsubscribe token?

Reply:
Ok!